lexos/processors/analyze/content_analysis_model.py

```diff
@@ -0,0 +1,461 @@
+        """Generates an html table from dataframe
+
+        :return: dataframe table in html
+        """
+        df = self.to_data_frame()
+        html = df.to_html(classes="table table-striped table-bordered",
+                          index=False)
+        return html
+
+    def to_data_frame(self) -> pd.DataFrame:
+        """Generates a dataframe containing all values stored in this class
+
+        :return: a data frame containing all values stored in this class
+        members
+        """
+        columns = ['Document Name'] + [dictionary.label for dictionary in
+                                       self.get_active_dicts()] + \
+                  ['formula', 'total word count', 'score']
+        indices = [file.label for file in self._corpus] + ['averages']
+        df = pd.DataFrame(columns=range(len(columns)),
+                          index=range(len(indices)))
+        # adds row to df with: file label, count for each dict,
+        # formula result, total word count, score
+        for i, (file, formula, score, counter) in enumerate(
+            zip(self._corpus, self._formulas,
+                self._scores, self._counters)):
+            df.xs(i)[0] = file.label
+            for j, (count) in enumerate(counter):
+                df.xs(i)[j + 1] = count
+            df.xs(i)[j + 2] = formula
+            df.xs(i)[j + 3] = file.total_word_counts
+            df.xs(i)[j + 4] = score
+        # add a row to df with the average of each column
+        for j, (average) in enumerate(self._averages):
+            df.xs(i + 1)[j] = average
+        df.columns = columns
+        return df
+
+    def is_secure(self) -> bool:
+        """Checks if the formula is secure
+
+        :return: True if secure, false if not secure
+        """
+        formula = self._formula
+        allowed_input = ["[" + dictionary.label + "]" for
+                         dictionary in self.get_active_dicts()] + \
+                        ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9",
+                         " ", "+", "-", "*", "/", "sin", "cos", "tan", "log",
+                         "sqrt", "(", ")"]
+        for item in allowed_input:
+            formula = formula.replace(item, "")
+        if len(formula) == 0:
+            return True
+        return False
+
+    def save_formula(self):
+        """saves formula from front-end in the class member _formula"""
+        if self._test_options is not None:
+            formula = self._test_options.formula
+        else:
+            formula = self.content_analysis_option.formula
+        if len(formula) == 0:
+            self._formula = "0"
+        else:
+            formula = formula.replace("√", "sqrt").replace("^", "**")
+            self._formula = formula
+        return self.check_formula()
+
+    def check_formula(self) -> str:
+        """Checks if there are any errors in the formula
+
+        :return: error message if there is an empty str or None if there is
```

Comment:
If you were just checking for an empty string, this would be a very short function. Please describe what conditions actually trigger an error message

Reply:
as the error messages say: it checks for mismatched parenthesis, and correct number of arguments for all allowed functions